core/lua: introduce func_adapter for Lua

drewdzzz opened the pull request:
We are going to introduce universal triggers registry and use it for
tarantool triggers (e.g. space.on_replace). However, when trigger is set
to the registry, we do not know which arguments will be passed there
from C. That is why current implementation of triggers in tarantool is not
suitable for use in universal registry.

We decided to introduce func_adapter interface which allows to call
functions from any programming language. Now we support only one
language, which is Lua, so for the most part this interface is needed
to invert dependencies.

The patch introduces func_adapter interface and its implementation for
Lua functions.

Part of https://github.com/tarantool/tarantool/issues/8657

drewdzzz commented on src/box/lua/func_adapter.h:
```diff
@@ -0,0 +1,26 @@
+ * registry. Never returns NULL.
+ */
+struct func_adapter *
+lua_func_adapter_create(int func_ref);
```
Done.

drewdzzz commented on src/box/lua/func_adapter.h:
```diff
@@ -0,0 +1,26 @@
+#endif /* defined(__cplusplus) */
+
+struct lua_State;
+struct lua_func_adapter;
```
Done.

drewdzzz commented on src/box/lua/func_adapter.c:
```diff
@@ -0,0 +1,343 @@
+		ctx->L = fiber()->storage.lua.stack;
+		ctx->coro_ref = LUA_REFNIL;
+	}
+	ctx->top_svp = lua_gettop(ctx->L);
```
Initialize it to 0 now.

drewdzzz commented on src/lib/core/func_adapter.h:
```diff
@@ -0,0 +1,309 @@
+static inline void
+func_adapter_end(struct func_adapter *func, struct func_adapter_ctx *ctx)
+{
+	func->vtab->end(ctx);
```
Done.

drewdzzz commented on src/box/lua/func_adapter.c:
```diff
@@ -0,0 +1,343 @@
+		(struct lua_func_adapter_ctx *)base_ctx;
+	int nargs = lua_gettop(ctx->L) - ctx->top_svp;
+	lua_rawgeti(ctx->L, LUA_REGISTRYINDEX, func->func_ref);
+	lua_insert(ctx->L, ctx->top_svp + 1);
```
Done.
After this change I pass `func_adapter` to `begin` instead of `call`.

drewdzzz commented on src/lib/core/func_adapter.h:
```diff
@@ -0,0 +1,309 @@
+	/**
+	 * Check if the next returned value is a tuple.
+	 */
+	bool (*is_tuple)(struct func_adapter_ctx *ctx);
```
Done.

drewdzzz commented on src/box/lua/func_adapter.c:
```diff
@@ -0,0 +1,343 @@
+/**
+ * Pops tuple. Returned tuple is referenced by the function and caller must
+ * unreference it.
+ */
```
Done.

drewdzzz commented on test/unit/lua_func_adapter.c:
```diff
@@ -0,0 +1,398 @@
+	for (int i = 0; i < 3; ++i) {
+		if (test_state->current > test_state->limit)
+			break;
+		func_adapter_push_integer(func, ctx, test_state->current++);
```
There is a plan to migrate all the triggers, including transactional ones, to the new registry. But transactional triggers will be migrated later (in the scope of follow ups), so I removed iterators from the patch - let's design them when we need it.

drewdzzz commented on src/lib/core/func_adapter.h:
```diff
@@ -0,0 +1,309 @@
+	/**
+	 * Pushes string argument.
+	 */
+	void (*push_lstring)(struct func_adapter_ctx *ctx, const char *str,
```
I renamed this virtual method to `push_string`. But what's about wrappers, we will need both of them - should I leave `func_adapter_push_lstring` wrapper as it is, or come up with a better name? (Maybe, `push_cstring` for zero terminated strings and `push_string` for strings with explicitly passed length?)
As an alternative, we can always pass length (explicitly pass `strlen(str)` for zero-terminated ones).

drewdzzz replied:
Thanks, named it so.

locker commented on src/lib/core/func_adapter.h:
```diff
@@ -0,0 +1,284 @@
+
+static inline void
+func_adapter_push_string(struct func_adapter *func,
+			 struct func_adapter_ctx *ctx, const char *str)
```
Better call it `push_str0`. Also, there's no point to stick to Lua API naming conventions. Better name the methods `push_str`, `push_str0`, `push_int`, `push_double`. It'd be more consistent with other parts of the Tarantool internals.

drewdzzz replied:
Done (`push_int` was removed after f2f discussion).

locker commented on src/lib/core/func_adapter.h:
```diff
@@ -0,0 +1,284 @@
+	/**
+	 * Check if the next returned value is a number.
+	 */
+	bool (*is_number)(struct func_adapter_ctx *ctx);
```
Does it return true if the value is of type `int64_t`? If yes, what would `pop_number` return in this case? Can it result in loss of precision? (e.g. 2^63-1 can't be stored in double precisely). Similar question to `is_integer`: does it return true if the value has a fractional part?

Please describe the behavior in the comment.

drewdzzz replied:
Left such a comment:
```c
/**
  * Checks if the next returned value is a number that can be represented
  * by double without loss of precision.
  */
bool (*is_double)(struct func_adapter_ctx *ctx);
```

locker commented on src/box/lua/func_adapter.c:
```diff
@@ -0,0 +1,280 @@
+/**
+ * Specialization of func_adapter for Lua functions.
+ */
+struct lua_func_adapter {
```
`lua_` is the Lua API namespace. We shouldn't use it for Tarantool internals. Probably better call this `func_adapter_lua` (similarly to `port_lua`, `func_lua`).

drewdzzz replied:
Done. Note that the unit test is still called `lua_func_adapter.c` in order to group all unit tests for Lua.

locker commented on src/box/lua/func_adapter.c:
```diff
@@ -0,0 +1,280 @@
+lua_func_adapter_is_string(struct func_adapter_ctx *base)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	return lua_gettop(ctx->L) >= ctx->idx && lua_isstring(ctx->L, ctx->idx);
```
`lua_isstring` returns true for numbers, which is probably not what we want.

drewdzzz replied:
Thanks, changed it to `lua_type` here.

locker commented on src/box/lua/func_adapter.c:
```diff
@@ -0,0 +1,280 @@
+static bool
+lua_func_adapter_is_integer(struct func_adapter_ctx *base)
+{
+	/* There is no integer type in Lua. */
```
There's int64 type in Tarantool. Also, we could return true if the Lua number can be converted to `int64_t` and false otherwise.

drewdzzz replied:
Discussed f2f, decided not to imlement this functionality.

locker commented on test/unit/lua_func_adapter.c:
```diff
@@ -0,0 +1,267 @@
+	   "Returned number must be as expected");
+	is(integer_retval, integer_expected,
+	   "Returned integer must be as expected");
+
```
You should reset the Lua stack after each test (`lua_settop(L, 0)`) otherwise they might affect each other.

drewdzzz replied:
Done.

locker commented on test/unit/lua_func_adapter.c:
```diff
@@ -0,0 +1,267 @@
+#undef EPS
+
+static struct lua_State *L;
+static struct func_adapter_ctx ctx;
```
Nit: There's no need for `ctx` to be global. Please define it on stack where it's used.

drewdzzz replied:
Done.

locker commented on test/unit/lua_func_adapter.c:
```diff
@@ -0,0 +1,267 @@
+	tuple_init(NULL);
+
+	L = luaT_newteststate();
+	tarantool_L = L;
```
Nit: There's no need to make `L` global if you set `tarantool_L` to it anyway. Please use `tarantool_L` instead of `L`.

drewdzzz replied:
Done.

drewdzzz commented:
1. Since we pass a Lua object directly, I added a test case that checks if a callable table can be wrapped by a `func_adapter`.
2. I moved typecheck assertions to virtual method wrappers, an example:
```c
static inline void
func_adapter_pop_double(struct func_adapter *func,
			struct func_adapter_ctx *ctx, double *out)
{
	assert(func_adapter_is_double(func, ctx));
	func->vtab->pop_double(ctx, out);
}
```
